brain/synapse.cpp

```diff
@@ -0,0 +1,175 @@
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+
+#include "synapse.h"
+#include "synapses.h"
+
+#include <vmmlib/vector.hpp>
+
+namespace brain
+{
+
+struct Synapse::Impl
+{
+    Impl( const Synapses& synapses, const size_t index )
+        : _synapses( synapses )
+        , _index( index )
+    {}
+
+    const Synapses& _synapses;
+    const size_t _index;
+};
+
+Synapse::Synapse( const Synapses& synapses, const size_t index )
+    : _impl( new Impl( synapses, index ))
+{
+}
+
+Synapse::~Synapse()
+{
+}
+
+Synapse::Synapse( const Synapse& rhs )
+    : _impl( new Impl( rhs._impl->_synapses, rhs._impl->_index ))
+{
+}
+
+Synapse::Synapse( Synapse&& rhs )
+    : _impl( std::move( rhs._impl ))
+{
+}
+
+Synapse& Synapse::operator=( Synapse&& rhs )
+{
+    if( this != &rhs )
+        _impl = std::move( rhs._impl );
+    return *this;
+}
+
+SynapseGID Synapse::getGID() const
+{
+    return std::make_pair( getPostsynapticGID(),
+                           _impl->_synapses.index()[_impl->_index] );
+}
+
+uint32_t Synapse::getPresynapticGID() const
+{
+    return _impl->_synapses.preGID()[_impl->_index];
+}
+
+uint32_t Synapse::getPresynapticSectionID() const
+{
+    return _impl->_synapses.preSectionID()[_impl->_index];
+}
+
+uint32_t Synapse::getPresynapticSegmentID() const
+{
+    return _impl->_synapses.preSegmentID()[_impl->_index];
+}
+
+float Synapse::getPresynapticDistance() const
+{
+    return _impl->_synapses.preDistance()[_impl->_index];
+}
+
+Vector3f Synapse::getPresynapticSurfacePosition() const
+{
+    return Vector3f( _impl->_synapses.preSurfacePositionX()[_impl->_index],
+                     _impl->_synapses.preSurfacePositionY()[_impl->_index],
+                     _impl->_synapses.preSurfacePositionZ()[_impl->_index] );
+}
+
+Vector3f Synapse::getPresynapticCenterPosition() const
+{
+    return Vector3f( _impl->_synapses.preCenterPositionX()[_impl->_index],
+                     _impl->_synapses.preCenterPositionY()[_impl->_index],
+                     _impl->_synapses.preCenterPositionZ()[_impl->_index] );
+}
+
+uint32_t Synapse::getPostsynapticGID() const
+{
+    return _impl->_synapses.postGID()[_impl->_index];
+}
+
+uint32_t Synapse::getPostsynapticSectionID() const
+{
+    return _impl->_synapses.postSectionID()[_impl->_index];
+}
+
+uint32_t Synapse::getPostsynapticSegmentID() const
+{
+    return _impl->_synapses.postSegmentID()[_impl->_index];
+}
+
+float Synapse::getPostsynapticDistance() const
+{
+    return _impl->_synapses.postDistance()[_impl->_index];
+}
+
+Vector3f Synapse::getPostsynapticSurfacePosition() const
+{
+    return Vector3f( _impl->_synapses.postSurfacePositionX()[_impl->_index],
+                     _impl->_synapses.postSurfacePositionY()[_impl->_index],
+                     _impl->_synapses.postSurfacePositionZ()[_impl->_index] );
+}
```

Comment:
In BBPSDK the positions where computed if the data file was not available. Should we do the same or can we safely assume that all circuits have position files (that was not true before)?
In some cases it might be cheaper to compute the position rather than loading it, but that requires loading the morphology (both pre and post for surface positions) and it only makes sense if the pre post section, segment ids are needed for something else, otherwise it's actually better to simply load the position.
Maybe the best solution is to provide some free functions that are able to compute the positions if needed.